File: 08_function_calling/01_ollama.R
# 01_ollama.R - Launch Ollama from R
# Pairs with 01_ollama.sh. Starts ollama serve in the background with system2.
# Tim Fraser

# Configuration (match 01_ollama.sh)
PORT = 11434L
OLLAMA_HOST = paste0("0.0.0.0:", PORT)
OLLAMA_CONTEXT_LENGTH = 32000L

# Set env for this R process and any child processes (e.g. system2).
Sys.setenv(OLLAMA_HOST = OLLAMA_HOST, OLLAMA_CONTEXT_LENGTH = as.character(OLLAMA_CONTEXT_LENGTH))

# Start ollama serve in the background. wait = FALSE so R does not block.
# stdout/stderr go to null so the console is not flooded.
system2("ollama", "serve", stdout = FALSE, stderr = FALSE, wait = FALSE)

# Optional: give the server a moment to bind before using it.
Sys.sleep(1)

# To stop the server later (run in R or in a separate script):
#   system2("pkill", c("-f", "ollama serve"), stdout = FALSE, stderr = FALSE)
# Or on Windows: system("taskkill /F /IM ollama.exe", ignore.stderr = TRUE)

File: 07_rag/01_ollama.py
# 01_ollama.py
# Launch Ollama from Python
# Pairs with 01_ollama.sh / 01_ollama.R
# Tim Fraser (Python adaptation)

# This script configures environment variables for Ollama,
# then starts `ollama serve` in the background without blocking
# the Python session. Useful for starting a local LLM server
# from within Python notebooks or scripts.

import os
import subprocess
import time

# 0. Setup #################################

## 0.1 Configuration ############################

PORT = 11434  # Match 01_ollama.sh
OLLAMA_HOST = f"0.0.0.0:{PORT}"
OLLAMA_CONTEXT_LENGTH = 32000

# Set environment variables for this process and any child processes
os.environ["OLLAMA_HOST"] = OLLAMA_HOST
os.environ["OLLAMA_CONTEXT_LENGTH"] = str(OLLAMA_CONTEXT_LENGTH)

## 0.2 Start Ollama Server #######################

# Start `ollama serve` in the background.
# stdout/stderr are redirected to DEVNULL so the console is not flooded.
process = subprocess.Popen(
    ["ollama", "serve"],
    stdout=subprocess.DEVNULL,
    stderr=subprocess.DEVNULL,
)

# Give the server a moment to start listening before use
time.sleep(1)

# Optional: you can keep a reference to `process` if you want to stop it later
# For example:
# process.terminate()  # or process.kill()

# On Windows, from a separate shell you can also stop it with:
#   taskkill /F /IM ollama.exe